File: python/devin/sample-agent/agent.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

"""
MyAgent – Agent 365 sample that integrates with Devin AI.

This agent demonstrates how to:
- Receive messages and forward them to Devin AI
- Handle notifications (email, Teams, etc.)
- Return responses through the Agent 365 SDK
- Integrate with Agent 365 Observability
"""

import logging
from typing import Optional

from microsoft_agents.hosting.core import Authorization, TurnContext
from microsoft_agents_a365.notifications.agent_notification import (
    AgentNotificationActivity,
    NotificationTypes,
)

from client import get_client

logger = logging.getLogger(__name__)


class MyAgent:
    """
    Devin AI proxy agent.

    Implements the same interface expected by the Agent 365 hosting framework.
    """

    def __init__(self) -> None:
        self.logger = logging.getLogger(self.__class__.__name__)

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    async def initialize(self) -> None:
        """Called once by the host before the first request."""
        logger.info("Devin AI agent initialized")

    async def cleanup(self) -> None:
        """Called on server shutdown."""
        logger.info("Devin AI agent cleanup completed")

    # ------------------------------------------------------------------
    # Message handling
    # ------------------------------------------------------------------
    # NOTE: ``auth`` and ``auth_handler_name`` are accepted to match the
    # interface expected by the host (see other Python samples). The Devin
    # SDK uses its own ``DEVIN_SDK_API_KEY``, so this sample does not exchange
    # an A365 OBO token before calling Devin.

    async def process_user_message(
        self,
        message: str,
        auth: Authorization,
        auth_handler_name: Optional[str],
        context: TurnContext,
    ) -> str:
        """
        Forward *message* to the Devin AI agent and return its response.
        """
        # Log user identity – populated by the A365 platform on every turn.
        from_prop = context.activity.from_property
        logger.info(
            "Turn received from user — DisplayName: '%s', UserId: '%s', AadObjectId: '%s'",
            getattr(from_prop, "name", None) or "(unknown)",
            getattr(from_prop, "id", None) or "(unknown)",
            getattr(from_prop, "aad_object_id", None) or "(none)",
        )

        try:
            client = get_client()
            response = await client.invoke_inference_scope(message, context)
            return response
        except Exception:
            # Log full traceback for debugging; return a generic message so we
            # don't leak internal details (config, identifiers, upstream errors).
            logger.exception("Devin AI query error")
            return "Sorry, I had trouble reaching Devin. Please try again."

    # ------------------------------------------------------------------
    # Notification handling
    # ------------------------------------------------------------------

    async def handle_agent_notification_activity(
        self,
        notification_activity: AgentNotificationActivity,
        auth: Authorization,
        auth_handler_name: Optional[str],
        context: TurnContext,
    ) -> str:
        """
        Route agent notifications to the appropriate handler.
        """
        notification_type = notification_activity.notification_type
        logger.info("Processing notification: %s", notification_type)

        if notification_type == NotificationTypes.EMAIL_NOTIFICATION:
            return await self._handle_email_notification(
                notification_activity, auth, auth_handler_name, context
            )

        # Generic / unsupported notification types
        logger.info("Received notification of type: %s", notification_type)
        return f"Received notification of type: {notification_type}"

    # ------------------------------------------------------------------
    # Email notification
    # ------------------------------------------------------------------

    async def _handle_email_notification(
        self,
        activity: AgentNotificationActivity,
        auth: Authorization,
        auth_handler_name: Optional[str],
        context: TurnContext,
    ) -> str:
        """
        Handle email notifications by forwarding the email content to
        Devin AI and returning the response.
        """
        email = getattr(activity, "email", None)
        if not email:
            return "I could not find the email notification details."

        try:
            sender_name = (
                getattr(context.activity.from_property, "name", None)
                or "unknown sender"
            )
            email_id = getattr(email, "id", "")
            conversation_id = getattr(email, "conversation_id", "")

            email_prompt = (
                f"You have a new email from {sender_name} with id '{email_id}', "
                f"ConversationId '{conversation_id}'. "
                "Please process this email and provide a helpful response."
            )

            client = get_client()
            response = await client.invoke_inference_scope(
                email_prompt, context
            )
            return (
                response
                or "I have processed your email but do not have a response at this time."
            )
        except Exception:
            logger.exception("Email notification error")
            return "Unable to process your email at this time."

File: nodejs/devin/sample-agent/src/agent.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import {
  AgentDetails,
  BaggageBuilder,
  InferenceDetails,
  InferenceOperationType,
  InferenceScope,
  InvokeAgentScope,
  InvokeAgentScopeDetails,
  ObservabilityManager,
  Request,
  TenantDetails,
} from "@microsoft/agents-a365-observability";
import { ClusterCategory } from "@microsoft/agents-a365-runtime";
import { Activity, ActivityTypes } from "@microsoft/agents-activity";
import {
  AgentApplication,
  AgentApplicationOptions,
  MemoryStorage,
  TurnContext,
  TurnState,
} from "@microsoft/agents-hosting";
import '@microsoft/agents-a365-notifications';
import {
  AgentNotificationActivity,
  NotificationType,
  createEmailResponseActivity,
} from "@microsoft/agents-a365-notifications";
import { Stream } from "stream";
import { v4 as uuidv4 } from "uuid";
import { devinClient } from "./devin-client";
import tokenCache from "./token-cache";
import { ApplicationTurnState } from "./types/agent.types";
import { getAgentDetails, getTenantDetails } from "./utils";

export class A365Agent extends AgentApplication<ApplicationTurnState> {
  isApplicationInstalled: boolean = false;
  agentName = "Devin Agent";

  constructor(
    options?: Partial<AgentApplicationOptions<ApplicationTurnState>> | undefined
  ) {
    super(options);
    const clusterCategory: ClusterCategory =
      (process.env.CLUSTER_CATEGORY as ClusterCategory) || "dev";

    // Initialize Observability SDK
    const observabilitySDK = ObservabilityManager.configure((builder) =>
      builder
        .withService("devin-sample-agent", "1.0.0")
        .withTokenResolver(async (agentId, tenantId) => {
          // Token resolver for authentication with Agent 365 observability
          console.log(
            "🔑 Token resolver called for agent:",
            agentId,
            "tenant:",
            tenantId
          );

          // Retrieve the cached agentic token
          const cacheKey = this.createAgenticTokenCacheKey(agentId, tenantId);
          const cachedToken = tokenCache.get(cacheKey);

          if (cachedToken) {
            console.log("🔑 Token retrieved from cache successfully");
            return cachedToken;
          }

          console.log(
            "⚠️ No cached token found - token should be cached during agent invocation"
          );
          return null;
        })
        .withClusterCategory(clusterCategory)
    );

    // Start the observability SDK
    observabilitySDK.start();

    // Handle messages
    this.onActivity(
      ActivityTypes.Message,
      async (context: TurnContext, state: ApplicationTurnState) => {
        // Increment count state
        let count = state.conversation.count ?? 0;
        state.conversation.count = ++count;

        // Extract agent and tenant details from context
        const invokeAgentDetails = getAgentDetails(context);
        const tenantDetails = getTenantDetails(context);

        // Create BaggageBuilder scope
        const baggageScope = new BaggageBuilder()
          .tenantId(tenantDetails.tenantId)
          .agentId(invokeAgentDetails.agentId)
          .agentName(invokeAgentDetails.agentName)
          .conversationId(context.activity.conversation?.id)
          .build();

        await baggageScope.run(async () => {
          const request: Request = {
            conversationId: context.activity.conversation?.id,
            sessionId: context.activity.conversation?.id,
            content: context.activity.text || undefined,
          };
          const invokeScopeDetails: InvokeAgentScopeDetails = {};
          const invokeAgentScope = InvokeAgentScope.start(
            request,
            invokeScopeDetails,
            { ...invokeAgentDetails, tenantId: tenantDetails.tenantId }
          );

          await invokeAgentScope.withActiveSpanAsync(async () => {
            invokeAgentScope.recordInputMessages([
              context.activity.text ?? "Unknown text",
            ]);

            await this.handleAgentMessageActivity(
              context,
              invokeAgentScope,
              invokeAgentDetails,
              tenantDetails
            );
          });

          invokeAgentScope.dispose();
        });

        baggageScope.dispose();
      }
    );

    // Handle agent notifications
    this.onAgentNotification(
      "agents:*",
      async (
        context: TurnContext,
        state: ApplicationTurnState,
        agentNotificationActivity: AgentNotificationActivity
      ) => {
        await this.handleAgentNotificationActivity(
          context,
          state,
          agentNotificationActivity
        );
      }
    );

    // Handle installation activities
    this.onActivity(
      ActivityTypes.InstallationUpdate,
      async (context: TurnContext, state: TurnState) => {
        await this.handleInstallationUpdateActivity(context, state);
      }
    );
  }

  /**
   * Handles incoming user messages and sends responses.
   */
  async handleAgentMessageActivity(
    turnContext: TurnContext,
    invokeAgentScope: InvokeAgentScope,
    agentDetails: AgentDetails,
    tenantDetails: TenantDetails
  ): Promise<void> {
    if (!this.isApplicationInstalled) {
      await turnContext.sendActivity(
        "Please install the application before sending messages."
      );
      return;
    }

    const userMessage = turnContext.activity.text?.trim() || "";

    const from = turnContext.activity?.from;
    console.log(`Turn received from user — DisplayName: '${from?.name ?? "(unknown)"}', UserId: '${from?.id ?? "(unknown)"}', AadObjectId: '${from?.aadObjectId ?? "(none)"}'`);

    if (!userMessage) {
      await turnContext.sendActivity(
        "Please send me a message and I'll help you!"
      );
      return;
    }

    // Multiple messages pattern: send an immediate acknowledgment before the LLM work begins.
    // Each sendActivity call produces a discrete Teams message.
    // NOTE: For Teams agentic identities, streaming is buffered into a single message by the SDK;
    //       use sendActivity for any messages that must arrive immediately.
    
    await turnContext.sendActivity('Got it — working on it…');

    // Typing indicator loop — refreshes the "..." animation every ~4s for long-running operations.
    // Typing indicators time out after ~5s and must be re-sent. Only visible in 1:1 and small group chats.
    let typingInterval: ReturnType<typeof setInterval> | undefined;
    const startTypingLoop = () => {
      typingInterval = setInterval(() => {
        turnContext.sendActivity(Activity.fromObject({ type: "typing" })).catch(() => {
          // Typing indicator failed — non-critical, continue
        });
      }, 4000);
    };
    const stopTypingLoop = () => { clearInterval(typingInterval); };

    startTypingLoop();

    let inferenceScope!: ReturnType<typeof InferenceScope.start>;
    try {
      const inferenceDetails: InferenceDetails = {
        operationName: InferenceOperationType.CHAT,
        model: "claude-3-7-sonnet-20250219",
        providerName: "cognition-ai",
        inputTokens: Math.ceil(userMessage.length / 4), // Rough estimate
        outputTokens: 0, // Will be updated after response
        finishReasons: undefined,
      };

      inferenceScope = InferenceScope.start(
        { conversationId: turnContext.activity.conversation?.id },
        inferenceDetails,
        { ...agentDetails, tenantId: agentDetails.tenantId || tenantDetails.tenantId }
      );
      inferenceScope.recordInputMessages([userMessage]);

      const chunks: string[] = [];
      let streamErrorMessage: string | undefined;
      let totalResponseLength = 0;
      const responseStream = new Stream()
        .on("data", (chunk) => {
          const text = chunk as string;
          totalResponseLength += text.length;
          chunks.push(text);
          invokeAgentScope.recordOutputMessages([`LLM Response: ${text}`]);
          inferenceScope.recordOutputMessages([`LLM Response: ${text}`]);
        })
        .on("error", (error) => {
          streamErrorMessage = String(error);
          invokeAgentScope.recordOutputMessages([`Streaming error: ${error}`]);
          inferenceScope.recordOutputMessages([`Streaming error: ${error}`]);
        })
        .on("close", () => {
          inferenceScope.recordOutputTokens(Math.ceil(totalResponseLength / 4));
          inferenceScope.recordFinishReasons(["stop"]);
        });

      await devinClient.invokeAgent(userMessage, responseStream);
      stopTypingLoop();
      if (streamErrorMessage) {
        await turnContext.sendActivity("There was an error processing your request, please try again.");
      } else if (chunks.length > 0) {
        await turnContext.sendActivity(chunks.join("\n\n"));
      } else {
        await turnContext.sendActivity("Devin did not return a response. Please try again.");
      }
    } catch (error) {
      stopTypingLoop();
      invokeAgentScope.recordOutputMessages([`LLM error: ${error}`]);
      await turnContext.sendActivity(
        "There was an error processing your request"
      );
    } finally {
      inferenceScope?.dispose();
    }
  }

  /**
   * Handles agent notification activities.
   */
  async handleAgentNotificationActivity(
    context: TurnContext,
    state: ApplicationTurnState,
    agentNotificationActivity: AgentNotificationActivity
  ): Promise<void> {
    switch (agentNotificationActivity.notificationType) {
      case NotificationType.EmailNotification:
        await this.handleEmailNotification(context, agentNotificationActivity);
        break;
      default:
        await context.sendActivity(
          `Received notification of type: ${agentNotificationActivity.notificationType}`
        );
    }
  }

  /**
   * Handles email notification activities with proper EmailResponse.
   */
  private async handleEmailNotification(
    context: TurnContext,
    activity: AgentNotificationActivity
  ): Promise<void> {
    const emailNotification = activity.emailNotification;

    if (!emailNotification) {
      const errorResponse = createEmailResponseActivity(
        "I could not find the email notification details."
      );
      await context.sendActivity(errorResponse);
      return;
    }

    try {
      // Collect the response from Devin using a stream
      let responseContent = "";
      const responseStream = new Stream()
        .on("data", (chunk) => {
          responseContent += chunk as string;
        })
        .on("error", (error) => {
          console.error("Stream error:", error);
        });

      // Process the email notification with Devin
      const prompt = `You have a new email from ${context.activity.from?.name} with id '${emailNotification.id}', ` +
        `ConversationId '${emailNotification.conversationId}'. Please process this email and provide a helpful response.`;

      await devinClient.invokeAgent(prompt, responseStream);

      const emailResponseActivity = createEmailResponseActivity(
        responseContent || "I have processed your email but do not have a response at this time."
      );
      await context.sendActivity(emailResponseActivity);
    } catch (error) {
      console.error("Email notification error:", error);
      const errorResponse = createEmailResponseActivity(
        "Unable to process your email at this time."
      );
      await context.sendActivity(errorResponse);
    }
  }

  /**
   * Handles agent installation and removal events.
   */
  async handleInstallationUpdateActivity(
    turnContext: TurnContext,
    state: TurnState
  ): Promise<void> {
    if (turnContext.activity.action === "add") {
      this.isApplicationInstalled = true;
      await turnContext.sendActivity(
        "Thank you for hiring me! Looking forward to assisting you in your professional journey!"
      );
    } else if (turnContext.activity.action === "remove") {
      this.isApplicationInstalled = false;
      await turnContext.sendActivity(
        "Thank you for your time, I enjoyed working with you."
      );
    }
  }

  /**
   * Create a cache key for the agentic token
   */
  private createAgenticTokenCacheKey(
    agentId: string,
    tenantId: string
  ): string {
    return tenantId
      ? `agentic-token-${agentId}-${tenantId}`
      : `agentic-token-${agentId}`;
  }
}

export const agentApplication = new A365Agent({
  storage: new MemoryStorage(),
  authorization: { agentic: {} }, // Type and scopes set in .env
});
